### PR TITLE
fix kubelet flushlogs not working together with exit

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
kubelet flushlogs not working together with exit

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/102231

#### Does this PR introduce a user-facing change?

```release-note
fix failed flushing logs in defer function when kubelet cmd exit 1.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by endocrimes** [comment]
/triage accepted
/priority important-longterm

@kerthcet Might be worth adding a release note entry as this might be useful to folks debugging deployments.

**Review by sanposhiho — COMMENTED** [review]
Thanks for helping the issue!
Also, in this PR, please check the below point. 

https://github.com/kubernetes/kubernetes/pull/104775#discussion_r704792673

On this PR, you have to consider, for example, this point
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L160

**Comment by kerthcet** [comment]
> Thanks for helping the issue!
> Also, in this PR, please check the below point.
> 
> [#104775 (comment)](https://github.com/kubernetes/kubernetes/pull/104775#discussion_r704792673)
> 
> On this PR, you have to consider, for example, this point
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L160

done :)

File: cmd/kubelet/kubelet.go
Comment (mrunalp) [inline cmd/kubelet/kubelet.go:39]: Do we want to log the error before we exit?

Comment (kerthcet) [reply cmd/kubelet/kubelet.go:39]: Refer to https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/kubelet.go#L40-L45, we want to flush the logs whether we exit or not, however, if  `command.Execute()` returns with error, defer func will not call.

Comment (pacoxu) [reply cmd/kubelet/kubelet.go:39]: /lgtm

https://github.com/kubernetes/kubernetes/blob/b18d6ee61d0c7e54ddadd7bb5351006ca840c80c/vendor/github.com/spf13/cobra/command.go#L956-L960
For cobra command,  there is an option `SilenceErrors` which is false by default(in kubelet we don't change it.)
In this case, the error will be printed inside `command.Execute()` so we do not have to log the error here.

**Review by sanposhiho — COMMENTED** [review]
plz check this for this pr too 🙏 
https://github.com/kubernetes/kubernetes/pull/104776#discussion_r716367629

File: cmd/kubelet/kubelet.go
Comment (kerthcet) [reply cmd/kubelet/kubelet.go:39]: logged already in https://github.com/kubernetes/kubernetes/pull/104774/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bL158

**Review by sanposhiho — COMMENTED** [review]
Thanks :)
/lgtm

File: cmd/kubelet/kubelet.go
Comment (sanposhiho) [inline cmd/kubelet/kubelet.go:38]: I think we don't have to use defer here.
- whenever runKubeletCmd() returns error, we want to log and exit(1)
- when runKubeletCmd() panic, we can see stack trace (as you said ([ref](https://github.com/kubernetes/kubernetes/pull/104776#discussion_r715827949))) and I agree with your opinion -- we don't have to recover and print it, I think.

Are there any cases where this is preferable?

**Review by sanposhiho — COMMENTED** [review]
Sorry, one more thing.
/hold

File: cmd/kubelet/kubelet.go
Comment (sanposhiho) [reply cmd/kubelet/kubelet.go:39]: maybe @mrunalp want to say: Do you forgot to print error here?
like your other PR
https://github.com/kubernetes/kubernetes/pull/104776/files#diff-dc7793108c9bdf4cd3d65595577d37c4ff125a40d5a162fc9cb89f69b294931cR36

Comment (kerthcet) [reply cmd/kubelet/kubelet.go:39]: Logs are printed in func `RunE`:
https://github.com/kubernetes/kubernetes/pull/104774/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bL158
https://github.com/kubernetes/kubernetes/pull/104774/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bL158
https://github.com/kubernetes/kubernetes/pull/104774/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bL158

I didn't change former behavior  as logs differed with each other.

Comment (sanposhiho) [reply cmd/kubelet/kubelet.go:39]: Aha, OK. Understood 👍

File: cmd/kubelet/kubelet.go
Comment (kerthcet) [reply cmd/kubelet/kubelet.go:38]: since no recover() will called, I'll refactor this code. the same with others.

**Review by sanposhiho — COMMENTED** [review]
Thanks :)

/lgtm
/hold cancel

**Comment by pacoxu** [comment]
/lgtm
/assign mrunalp
to take a look again.

File: cmd/kubelet/app/server.go
Comment (kerthcet) [inline cmd/kubelet/app/server.go:300]: add `SilenceUsage=true` or usage would be printed twice. See https://github.com/kubernetes/kubernetes/pull/104774/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bR159

Comment (pacoxu) [reply cmd/kubelet/app/server.go:300]: /lgtm
Nice catch

File: cmd/kubelet/kubelet.go
Comment (pohly) [inline cmd/kubelet/kubelet.go:53]: https://github.com/kubernetes/kubernetes/pull/105076/files has a similar fix that makes `main` a bit more consistent with the way how other commands are set up (command constructed in main, executed in helper function).

It's a minor difference, but the two PRs will conflict.

Can we perhaps merge the other PR first (it touches many different areas and will be harder to get merged when rebasing is needed)?

Comment (kerthcet) [reply cmd/kubelet/kubelet.go:53]: of course, waiting for good news.

Comment (pohly) [reply cmd/kubelet/kubelet.go:53]: /cancel hold

The other PR was merged, please rebase. Thanks for your patience.

Comment (pohly) [reply cmd/kubelet/kubelet.go:53]: /hold cancel

Comment (kerthcet) [reply cmd/kubelet/kubelet.go:53]: thx

**Comment by kerthcet** [comment]
PTAL @pohly  @sanposhiho , just remove the `os.Exit()`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:170]: Nit, not worth fixing: this isn't a format string, so `errors.New` is more appropriate.

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:191]: Why this change?

Starting a log messages is [good practice](https://github.com/kubernetes/community/blob/HEAD/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message) - I know, it's the opposite of the best practice for error messages :sweat_smile:

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:204]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:208]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:235]: `err` is not getting wrapped here and `dynamicConfigDir` looks odd when embedded inside an error message like this.

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:243]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:260]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:187]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:182]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (pohly) [inline cmd/kubelet/app/server.go:158]: Replace `,` with `:`

File: cmd/kubelet/app/server.go
Comment (kerthcet) [reply cmd/kubelet/app/server.go]: sorry, missed the error msg, Done:)

File: cmd/kubelet/app/server.go
Comment (kerthcet) [reply cmd/kubelet/app/server.go]: thanks a lot for your patience. Done :)

**Review by pohly — APPROVED** [review]
Looks good to me, but I'll defer to the code owners for the final LGTM + approval.

**Comment by kerthcet** [comment]
kindly ping @endocrimes , @sanposhiho

**Comment by kerthcet** [comment]
No changes applied, just rebase the master branch.
/remove-kind bug
/kind feature

File: cmd/kubelet/app/server.go
Comment (ehashman) [inline cmd/kubelet/app/server.go:156]: https://pkg.go.dev/github.com/spf13/cobra#Command.RunE

File: cmd/kubelet/app/server.go
Comment (ehashman) [inline cmd/kubelet/app/server.go:160]: After this change we no longer print usage info. I suspect this is okay because the kubelet usage info is massively long and makes it hard to see the error, but we should note this in the release note.

File: cmd/kubelet/app/server.go
Comment (ehashman) [inline cmd/kubelet/app/server.go:159]: I think we typically prefer `%w` for errors to ensure they are unwrapped: https://pkg.go.dev/fmt#Errorf

**Review by ehashman — COMMENTED** [review]
Please update your first comment, this does not fix entirely fix the linked issue and GitHub will auto-close it if this merges.

File: cmd/kubelet/app/server.go
Comment (kerthcet) [reply cmd/kubelet/app/server.go:160]: We will still print usage info [here](https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/spf13/cobra/command.go#L991-L993)

File: cmd/kubelet/app/server.go
Comment (kerthcet) [reply cmd/kubelet/app/server.go:159]: done:)

**Comment by kerthcet** [comment]
kindly ping @mrunalp

**Comment by dims** [comment]
/assign @sjenning @klueska 

NOTE: there are 3 people in favor of this change (lgtm got lost!)

**Pull request merged**
